File: CharityDonation/backend/app/routers/donations.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.deps import get_current_user, get_db
from app.models.enums import ItemCategory
from app.models.user import User
from app.schemas.donation import (
    DonationBrowseItem,
    DonationCreate,
    DonationCreateResponse,
    DonationDetailResponse,
    DonationResponse,
    MatchPreviewItem,
)
from app.schemas.route import RouteResponse
from app.services import donation_service, route_service

router = APIRouter(prefix="/donations", tags=["donations"])


@router.post("", response_model=DonationCreateResponse, status_code=201)
def create_donation(
    payload: DonationCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return donation_service.create_donation(db, current_user, payload)


@router.get("", response_model=list[DonationResponse])
def list_donations(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return donation_service.list_donations(db, current_user)


@router.get("/all", response_model=list[DonationBrowseItem])
def list_all_donations(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return donation_service.list_all_donations(db)


@router.get("/matches", response_model=list[MatchPreviewItem])
def preview_matches(
    item_name: str = Query(...),
    item_category: ItemCategory = Query(...),
    lat: float = Query(...),
    lng: float = Query(...),
    description: str | None = Query(None),
    quantity: int = Query(1),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return donation_service.preview_matches(
        db, item_name, item_category, description, quantity, lat, lng,
        exclude_user_id=current_user.id,
    )


@router.get("/{donation_id}", response_model=DonationDetailResponse)
def get_donation(
    donation_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return donation_service.get_donation(db, current_user, donation_id)


@router.get("/{donation_id}/route", response_model=RouteResponse)
def get_donation_route(
    donation_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return route_service.get_pickup_route(db, current_user, donation_id)

File: CharityDonation/frontend/src/api/donations.ts
import { apiRequest } from "./client";
import type { Donation, Location, MatchPreview } from "../types";

export interface CreateDonationPayload {
  item_name: string;
  item_category: string;
  description?: string;
  quantity: number;
  pickup_location: Location;
  target_receiver_id?: string | null;
}

export function createDonation(payload: CreateDonationPayload) {
  return apiRequest<Donation>("/donations", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function listDonations() {
  return apiRequest<Donation[]>("/donations");
}

export function getDonation(id: string) {
  return apiRequest<Donation>(`/donations/${id}`);
}

export function previewMatches(params: {
  item_name: string;
  item_category: string;
  description?: string;
  lat: number;
  lng: number;
}) {
  const query = new URLSearchParams({
    item_name: params.item_name,
    item_category: params.item_category,
    lat: String(params.lat),
    lng: String(params.lng),
  });
  if (params.description) {
    query.set("description", params.description);
  }
  return apiRequest<MatchPreview[]>(`/donations/matches?${query.toString()}`);
}
